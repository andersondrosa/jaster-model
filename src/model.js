import { get, isObject, order, clone, nested, emptyObject } from "./utils.js";
import { merge, diff, sha256 } from "jaster";

function setStartData($this, data) {
  $this.startData = $this.commitData = JSON.stringify(order(data));
  $this.data = JSON.parse($this.startData);
}

function setCommitData($this) {
  $this.commitData = isObject($this.data)
    ? JSON.stringify(order($this.data))
    : "{}";
  $this.dirty = false;
}

export default class {
  //
  constructor(data, commits) {
    //
    this.data = {};
    this.startData = "{}";
    this.commitData = "{}";
    this.undoCommits = [];
    this.commits = [];
    this.dirty = false;

    this.prevData = [];
    this.nextData = [];

    if (data && isObject(data)) {
      setStartData(this, data);
    }

    if (Array.isArray(commits)) {
      this.setCommits(commits, true);
    }
  }

  /**
   * Setando os dados iniciais que serão usados de base caso precise restaurar
   * ou desfazer algum comit, por é necessário usar a base e re mesclar todos
   * os comits sobre a base
   */

  // ===========================================================================

  // GETTERS

  hasCommits() {
    return this.commits.length > 0;
  }

  get(key) {
    return get(key, this.data);
  }

  getCommits() {
    return this.commits.map(row => clone(row));
  }

  getCommitData() {
    if (!this.commitData) {
      return {};
    }
    return JSON.parse(this.commitData);
  }

  discardChanges() {
    //
    this.data = this.getCommitData();

    this.cleanHistoric();
  }

  getData() {
    if (this.data) {
      return order(this.data);
    }
    return {};
  }

  exportData() {
    return {
      data: JSON.parse(this.startData),
      commits: this.getCommits()
    };
  }

  getIntegrity() {
    return sha256(this.commitData);
  }

  getUncommitedData() {
    //
    if (this.dirty == false) {
      return {};
    }

    return order(diff(this.getCommitData(), this.data));
  }

  commitRequired() {
    //
    if (this.dirty == false) {
      return false;
    }

    if (this.commitData == JSON.stringify(this.getData())) {
      this.dirty = false;
    }

    return this.dirty;
  }

  // ===========================================================================
  // ===========================================================================

  // ACTIONS

  setCommits(commits) {
    this.commits = clone(commits);
    this.restore();
  }

  cleanHistoric() {
    this.prevData = [];
    this.nextData = [];
  }

  put(data) {
    //
    if (data instanceof Function) {
      data = data(this.data);
    }

    const info = merge(this.data, data);

    if (!info.unchanged) {
      this.prevData.push(info);
    }

    this.nextData = [];

    this.dirty = true;

    return this;
  }

  rollback(n = 1) {
    //
    for (let i = 0; i < n; i++) {
      const row = this.prevData.pop();

      if (row == undefined) {
        return false;
      }

      this.nextData.push(merge(this.data, row.rollback));

      this.dirty = true;
    }

    return !!this.dirty;
  }

  rollforth(n = 1) {
    //
    for (let i = 0; i < n; i++) {
      const row = this.nextData.pop();

      if (row == undefined) {
        return false;
      }

      this.prevData.push(merge(this.data, row.rollback));

      this.dirty = true;
    }

    return !!this.dirty;
  }

  // ===========================================================================
  // ===========================================================================

  validateIntegrity(fdata) {
    //
    let certified;

    if (fdata.hasOwnProperty("_certified")) {
      certified = fdata._certified;
      fdata._certified = "<hidden>";
    }

    const str_data = JSON.stringify(fdata);

    const hash = sha256(str_data);

    if (certified) {
      fdata._certified = certified;

      if ((certified === hash) == false) {
        throw "Invalid certified " +
          certified +
          ". Valid certified must match " +
          hash;
      }
    }
  }

  commit(message, type = "put") {
    //
    if (this.dirty == false) {
      return false;
    }

    const data = this.getUncommitedData();

    if (emptyObject(data)) {
      this.dirty = false;
      return false;
    }

    const commit = { type, message, data };

    if (!message) {
      delete commit.message;
    }

    this.commits.push(commit);

    const rollback = diff(this.data, this.getCommitData());

    this.undoCommits.push(rollback);

    setCommitData(this);

    return true;
  }

  cleanCommits() {
    //
    this.commits = [];

    this.restore();
  }

  rollbackCommit(levels = 1) {
    //
    this.data = this.getCommitData();

    for (let i = 0; i < levels; i++) {
      this.commits.pop();
      const undo = this.undoCommits.pop();
      merge(this.data, undo);
    }

    setCommitData(this);
  }

  // rollbackCommit(levels) {
  //   //
  //   const commits = [];

  //   levels = this.commits.length - levels;

  //   for (const i in this.commits) {
  //     if (i == levels) {
  //       this.commits = commits;
  //       break;
  //     }

  //     const commit = this.commits[i];

  //     commits.push(commit);
  //   }

  //   this.cleanHistoric();

  //   this.commits = commits;

  //   this.restore();
  // }

  restore() {
    //
    this.data = JSON.parse(this.startData);

    this.undoCommits = this.commits.map(({ data }) => {
      const { rollback } = merge(this.data, data);
      return rollback;
    });

    setCommitData(this);
  }

  // ===========================================================================
  // ===========================================================================
  // ===========================================================================

  // ALIAS

  putn(o) {
    return this.put(nested(o, true));
  }
}
