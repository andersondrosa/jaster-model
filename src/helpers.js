"use strict";

export function isObject(o) {
  return o instanceof Object && o.constructor.name == "Object";
}

export function emptyObject(o) {
  for (let i in o) {
    return false;
  }
  return true;
}

export function get(path, data) {
  let keys = path.split(".");

  let pointer = data;

  let _key = keys.pop();

  for (let i in keys) {
    let key = keys[i];

    if (typeof data != "object" || !pointer.hasOwnProperty(key)) {
      return;
    }

    pointer = pointer[key];
  }

  return pointer[_key] ? pointer[_key] : null;
}

export function lget(path, layers) {
  if (typeof layers != "object") {
    return;
  }

  if (path.indexOf("[") > -1) {
    path = path.replace(/\[([^\]]+)\]/g, (m, i) => {
      return "." + i;
    });
  }

  let i = path.indexOf(".");

  let root = i == -1 ? path : path.slice(0, i);

  let data;

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(root)) {
      data = layers[i][root];
      break;
    }
  }

  if (i == -1) {
    return data;
  }

  if (typeof data == "object") {
    return this.get(path.slice(i + 1), data);
  }
}

export function slget(key, layers, def) {
  if (typeof layers != "object") {
    return null;
  }

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(key)) {
      return layers[i][key];
    }
  }

  return def;
}

export function set(data, path, value) {
  let keys = path.split(".");

  let pointer = data;

  let _key = keys.pop();

  for (let i in keys) {
    let key = keys[i];

    if (!pointer.hasOwnProperty(key) || !isObject(pointer[key])) {
      pointer[key] = {};
    }

    pointer = pointer[key];
  }

  pointer[_key] = value;
}

export function parsePath(path, key) {
  if (path) {
    path = path.split(".");
  } else {
    path = [];
  }

  let keys = key.split("/");

  let realkey = keys.pop();

  for (let i in keys) {
    if (keys[i] == "root") {
      path = [];
      break;
    }

    if (keys[i] == "..") {
      path.pop();
    }
  }

  path.push(realkey);

  let realpath = path.join(".");

  return realpath;
}

export function size(object) {
  var length = 0;
  for (var key in object) {
    if (object.hasOwnProperty(key)) {
      ++length;
    }
  }
  return length;
}
