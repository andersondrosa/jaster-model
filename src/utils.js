"use strict";

export function isObject(o) {
  return o instanceof Object && o.constructor.name == "Object";
}

export function emptyObject(o) {
  for (let i in o) {
    return false;
  }
  return true;
}

export function get(path, data) {
  let keys = path.split(".");

  let pointer = data;

  let _key = keys.pop();

  for (let i in keys) {
    let key = keys[i];

    if (typeof data != "object" || !pointer.hasOwnProperty(key)) {
      return;
    }

    pointer = pointer[key];
  }

  return pointer[_key] ? pointer[_key] : null;
}

export function lget(path, layers) {
  if (typeof layers != "object") {
    return;
  }

  if (path.indexOf("[") > -1) {
    path = path.replace(/\[([^\]]+)\]/g, (m, i) => "." + i);
  }

  let i = path.indexOf(".");

  let root = i == -1 ? path : path.slice(0, i);

  let data;

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(root)) {
      data = layers[i][root];
      break;
    }
  }

  if (i == -1) {
    return data;
  }

  if (typeof data == "object") {
    return this.get(path.slice(i + 1), data);
  }
}

export function slget(key, layers, def) {
  if (typeof layers != "object") {
    return null;
  }

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(key)) {
      return layers[i][key];
    }
  }

  return def;
}

export function set(data, path, value) {
  const keys = path.split(".");
  let key;
  let pointer = data;
  let _key = keys.pop();
  for (let i in keys) {
    key = keys[i];
    if (!pointer.hasOwnProperty(key) || !isObject(pointer[key])) {
      pointer[key] = {};
    }
    pointer = pointer[key];
  }
  pointer[_key] = value;
}

export function parsePath(path, key) {
  if (path) {
    path = path.split(".");
  } else {
    path = [];
  }

  let keys = key.split("/");

  let realkey = keys.pop();

  for (let i in keys) {
    if (keys[i] == "root") {
      path = [];
      break;
    }

    if (keys[i] == "..") {
      path.pop();
    }
  }

  path.push(realkey);

  let realpath = path.join(".");

  return realpath;
}

export function size(object) {
  var length = 0;
  for (var key in object) {
    if (object.hasOwnProperty(key)) {
      ++length;
    }
  }
  return length;
}

function order(o) {
  const r = Array.isArray(o) ? [] : {};
  let v, k, i;
  const keys = Object.keys(o).sort();
  for (i in keys) {
    k = keys[i];
    v = o[k];
    r[k] = typeof v === "object" && v !== null ? order(v) : v;
  }
  return r;
}

function clone(o) {
  let r, v, k;
  if (typeof o !== "object" || o === null) {
    return o;
  }
  r = Array.isArray(o) ? [] : {};
  for (k in o) {
    v = o[k];
    r[k] = typeof v === "object" && v !== null ? clone(v) : v;
  }
  return r;
}

function nested(o, deep) {
  if (deep) {
    return deepNested(o);
  }
  const r = {};
  for (const k in o) {
    set(r, k, o[k]);
  }
  return r;
}

function deepNested(o) {
  //
  if (Array.isArray(o)) {
    return o.map(v => (v && typeof v == "object" ? deepNested(v) : v));
  }

  let v;
  const r = {};
  for (const k in o) {
    v = o[k];
    set(r, k, v && typeof v == "object" ? deepNested(v) : v);
  }
  return r;
}

export { order, clone, nested };
