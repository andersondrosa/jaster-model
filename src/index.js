import { get, isObject } from "./helpers.js";
import { merge, diff, sha256 } from "jaster";

export default class {
  //

  constructor(data, commits) {
    //
    this.data;
    this.baseData;
    this.commitData;

    this.state = {
      commits: []
    };

    this.prevData = [];
    this.nextData = [];

    this.undoCommits = [];

    if (isObject(data)) {
      this.setBaseData(data);
    }

    if (Array.isArray(commits)) {
      this.setCommits(commits, true);
    }
  }

  /**
   * Setando os dados iniciais que serão usados de base caso precise restaurar
   * ou desfazer algum comit, por é necessário usar a base e re mesclar todos
   * os comits sobre a base
   */
  setBaseData(data) {
    this.baseData = JSON.stringify(data);
    this.data = JSON.parse(this.baseData);
    // SE commitData for NULL, é para se usar o baseData
    this.commitData = null;
  }

  commitRequired() {
    return this.prevData.length > 0;
  }

  cleanHistoric(data) {
    this.prevData = [];
    this.nextData = [];
  }

  isPristine() {
    return !(this.state.commits.length > 0);
    // return JSON.stringify(this.data) == "{}" && this.prevData.length == 0;
  }

  put(data) {
    //
    if (this.data == null) {
      this.data = data;
      this.registerData();
      return this;
    }

    if (data instanceof Function) {
      data = data(this.data);
    }

    let info = merge(this.data, data);

    if (!info.unchanged) {
      this.prevData.push(info);
    }

    this.nextData = [];

    return this;
  }

  rollback(n = 1) {
    var l = this.prevData.length;

    for (let i = 0; i < n; i++) {
      const row = this.prevData.pop();

      if (row == undefined) {
        return;
      }

      const info = merge(this.data, row.rollback);

      this.nextData.push(info);
    }
  }

  rollforth(n = 1) {
    var l = this.nextData.length;

    for (let i = 0; i < n; i++) {
      const row = this.nextData.pop();

      if (row == undefined) {
        return;
      }

      const info = merge(this.data, row.rollback);

      this.prevData.push(info);
    }
  }

  get(path) {
    return get(path, this.data);
  }

  getPath(data, path) {
    //
    let paths = [];

    if (path) {
      for (const key in data) {
        if (data[key].constructor === Object) {
          paths = paths.concat(this.getPath(data[key], path + "." + key));
          continue;
        }

        paths.push(path + "." + key);
      }

      return paths;
    }

    for (const key in data) {
      if (data[key].constructor === Object) {
        paths = paths.concat(this.getPath(data[key], key));
        continue;
      }

      paths.push(key);
    }

    return paths;
  }

  // ===========================================================================
  // ===========================================================================

  validateIntegrity(fdata) {
    let certified;

    if (fdata.hasOwnProperty("_certified")) {
      certified = fdata._certified;
      fdata._certified = "<hidden>";
    }

    let str_data = JSON.stringify(fdata);

    let hash = sha256(str_data);

    if (certified) {
      fdata._certified = certified;

      if ((certified === hash) == false) {
        throw "Invalid certified " +
          certified +
          ". Valid certified must match " +
          hash;
      }
    }
  }

  getCommits() {
    return this.state.commits;
  }

  commit(message) {
    //
    let commitData = diff(this.getComparisonData(), this.data);

    this.addCommit("PUT", message, commitData);

    return this;
  }

  addCommit(type, message, data) {
    //
    const commit = { type, message, data };

    message == undefined && delete commit.message;

    this.state.commits.push(commit);

    this.cleanHistoric();

    this.registerData();
  }

  cleanCommits() {
    //
    this.state.commits = [];
  }

  async rollbackCommit(levels, cb) {
    //
    let state = this.state;
    let commits = [];

    levels = state.commits.length - levels;

    for (const i in state.commits) {
      if (i == levels) {
        state.commits = commits;
        return;
      }

      const commit = state.commits[i];

      commits.push(commit);
    }

    this.cleanHistoric();

    state.commits = commits;

    await this.restore();

    cb && cb();
  }

  registerData() {
    this.commitData = JSON.stringify(this.data);
  }

  getComparisonData() {
    if (!this._data) {
      this._data = JSON.parse(this.commitData);
    }
    return this._data;
  }

  getData(utils) {
    return JSON.parse(JSON.stringify(this.data));
  }

  async getExtensionData(ext) {
    //
    if (!this.$fdata.dependencyProvider) {
      throw "Dependency Provider must be defined";
    }

    let data = {};

    let fdata = await this.$fdata.loadResource(ext);

    if (fdata._extends) {
      data = await this.getExtensionData(fdata._extends);
    }

    if (fdata._commits) {
      for (let i in fdata._commits) {
        merge(data, fdata._commits[i].data);
      }
    }

    return data;
  }

  async restore() {
    //
    const state = this.state;

    this.data = JSON.parse(this.baseData);

    for (const i in state.commits) {
      //
      const commit = state.commits[i];

      this.undoCommits.push(merge(this.data, commit.data));
    }

    this.registerData();
  }

  commitRequired() {
    //
    let request = { response: false };

    this.emmit("commit-required", [request]);

    return request.response == true;
  }

  setCommits(commits, restore) {
    this.state.commits = commits;
    restore == true && this.restore();
  }
}
