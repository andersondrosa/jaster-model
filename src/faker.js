import { nested, order } from "./utils";
const lorem = [
  "alpha",
  "beta",
  "gamma",
  "delta",
  "zeta",
  "lambda",
  "pi",
  "sigma",
  "omega",
  "sampi"
]

function getRandom(arr) {
  return arr[rand(arr.length - 1)];
}
function rand(num) {
  return Math.floor(Math.random() * (num + 1));
}

function fakerValue() {
  return getRandom(lorem);
}

function randLine() {
  let line = [];
  const l = rand(4);
  for (let i = 0; i < l; i++) {
    line.push(getRandom(lorem));
  }
  if(line.length == 0) {
    return getRandom(lorem)
  }
  return line.join(".");
}

function reduce(res, values) {
  const num = values.shift();
  if (!(num > 0)) {
    return res;
  }
  let response = {};
  let i = 0;
  for (let key in res) {
    if (i >= num) break;
    i++;
    if (typeof res[key] == "object" && values.length > 0) {
      response[key] = reduce(res[key], values.slice());
    } else {
      response[key] = res[key];
    }
  }
  return response;
}

function faker() {
  //
  let res = {};

  for (let i = 0; i < 100000; i++) {
    res[randLine()] = fakerValue();
    getRandom(lorem);
  }

  res = nested(res);

  const response = {
    alpha: reduce(res.alpha, [3, 2, 5]),
    beta: reduce(res.beta, [3, 2, 5]),
  };

  return response;
}

export default faker;
