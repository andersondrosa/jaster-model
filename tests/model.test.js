import Model from "../src";
import { object, rm } from "../src";
import { order, nested, deepNested } from "../src/utils";

function print(o) {
  console.log(JSON.stringify(o, null, 2));
}
//
const data = {
  first: "ok"
};

const $john = new Model(data);

it("Put first data", function() {
  //
  $john.put({
    name: "John",
    age: 30
  });

  expect($john.getData()).toMatchSnapshot();
});

// PUT =======================================================================
it("Put John Doe", function() {
  //
  $john.put(data => ({
    name: data.name + " Doe"
  }));

  expect($john.getData()).toMatchSnapshot();

  expect($john.getUncommitedData()).toMatchSnapshot();
});

// COMMIT ====================================================================
it("First commit", function() {
  //
  $john.commit();

  expect($john.getCommits()).toMatchSnapshot();
});

// PUT =======================================================================
it("Put Jane Doe", function() {
  //
  $john.put({ name: "Jane Doe" });

  expect($john.getData()).toMatchSnapshot();
});

// ROLLBACK ==================================================================
it("Rollback", function() {
  //
  $john.rollback();

  expect($john.getData()).toMatchSnapshot();
});

// ROLLFORTH==================================================================
it("Rollforth", function() {
  //
  $john.rollforth();

  expect($john.getData()).toMatchSnapshot();
});

// PUT =======================================================================
it("Put Mary Doe", function() {
  //
  $john.put({ name: "Mary Doe" });

  expect($john.getData()).toMatchSnapshot();

  expect($john.getUncommitedData()).toMatchSnapshot();
});

// ROLLBACK ==================================================================
it("Rollback Mary Doe", function() {
  //
  $john.rollback();

  expect($john.getData()).toMatchSnapshot();

  expect($john.getUncommitedData()).toMatchSnapshot();
});

// COMMIT ====================================================================
it("Second Commit", function() {
  //
  $john.commit();

  expect($john.getUncommitedData()).toMatchSnapshot();

  expect($john.getCommits()).toMatchSnapshot();

  expect($john.getData()).toMatchSnapshot();
});

// ROLLBACKL COMMIT ==========================================================
it("Rollback second commit", function() {
  //
  $john.rollbackCommit(1);

  expect($john.getCommits()).toMatchSnapshot();

  expect($john.getData()).toMatchSnapshot();

  expect($john.getUncommitedData()).toMatchSnapshot();
});

// ===========================================================
it("Final", function() {
  //
  let commits = $john.getCommits();

  const $mary = new Model(data, commits);

  $mary.put({
    alpha: "foo"
  });

  $mary.commit();

  $mary.put({
    omega: "baz",
    alpha: "bar"
  });

  $mary.commit();

  expect(commits).toMatchSnapshot();

  // print($mary.getCommits());

  expect($mary.getCommits()).toMatchSnapshot();

  $mary.restore();

  // print($mary.getCommits());

  expect($mary.undoCommits).toMatchSnapshot();

  $mary.rollbackCommit();
  expect($mary.undoCommits).toMatchSnapshot();
});

it("Test", function() {
  const $model = new Model({});

  $model.put({
    a: "1"
  });

  $model.commit("commit a:1");

  $model.put({
    b: "1"
  });

  $model.commit("commit b:1");

  $model.put({
    c: "1"
  });

  $model.putn({
    "d.b": "B",
    "d.a": "A"
  });

  $model.commit("commit c:1, d:1");

  $model.put({
    d: {
      ...rm("a"),
      a: "A"
    }
  });

  $model.commit();
  
  $model.rollback();
  $model.rollback();
  $model.rollforth();
  
  $model.commit();
  
  expect($model.getCommits()).toMatchSnapshot()
});
