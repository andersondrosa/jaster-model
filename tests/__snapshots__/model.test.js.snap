// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Final 1`] = `
Array [
  Object {
    "data": Object {
      "age": 30,
      "name": "John Doe",
    },
    "type": "put",
  },
]
`;

exports[`Final 2`] = `
Array [
  Object {
    "data": Object {
      "age": 30,
      "name": "John Doe",
    },
    "type": "put",
  },
  Object {
    "data": Object {
      "alpha": "foo",
    },
    "type": "put",
  },
  Object {
    "data": Object {
      "alpha": "bar",
      "omega": "baz",
    },
    "type": "put",
  },
]
`;

exports[`Final 3`] = `
Array [
  Object {
    "__rm": Array [
      "age",
      "name",
    ],
  },
  Object {
    "__rm": Array [
      "alpha",
    ],
  },
  Object {
    "__rm": Array [
      "omega",
    ],
    "alpha": "foo",
  },
]
`;

exports[`Final 4`] = `
Array [
  Object {
    "__rm": Array [
      "age",
      "name",
    ],
  },
  Object {
    "__rm": Array [
      "alpha",
    ],
  },
]
`;

exports[`First commit 1`] = `
Array [
  Object {
    "data": Object {
      "age": 30,
      "name": "John Doe",
    },
    "type": "put",
  },
]
`;

exports[`Put Jane Doe 1`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "Jane Doe",
}
`;

exports[`Put John Doe 1`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "John Doe",
}
`;

exports[`Put John Doe 2`] = `
Object {
  "age": 30,
  "name": "John Doe",
}
`;

exports[`Put Mary Doe 1`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "Mary Doe",
}
`;

exports[`Put Mary Doe 2`] = `
Object {
  "name": "Mary Doe",
}
`;

exports[`Put first data 1`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "John",
}
`;

exports[`Rollback 1`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "John Doe",
}
`;

exports[`Rollback Mary Doe 1`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "Jane Doe",
}
`;

exports[`Rollback Mary Doe 2`] = `
Object {
  "name": "Jane Doe",
}
`;

exports[`Rollback second commit 1`] = `
Array [
  Object {
    "data": Object {
      "age": 30,
      "name": "John Doe",
    },
    "type": "put",
  },
]
`;

exports[`Rollback second commit 2`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "John Doe",
}
`;

exports[`Rollback second commit 3`] = `Object {}`;

exports[`Rollforth 1`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "Jane Doe",
}
`;

exports[`Second Commit 1`] = `Object {}`;

exports[`Second Commit 2`] = `
Array [
  Object {
    "data": Object {
      "age": 30,
      "name": "John Doe",
    },
    "type": "put",
  },
  Object {
    "data": Object {
      "name": "Jane Doe",
    },
    "type": "put",
  },
]
`;

exports[`Second Commit 3`] = `
Object {
  "age": 30,
  "first": "ok",
  "name": "Jane Doe",
}
`;

exports[`Test 1`] = `
Array [
  Object {
    "data": Object {
      "a": "1",
    },
    "message": "commit a:1",
    "type": "put",
  },
  Object {
    "data": Object {
      "b": "1",
    },
    "message": "commit b:1",
    "type": "put",
  },
  Object {
    "data": Object {
      "c": "1",
      "d": Object {
        "a": "A",
        "b": "B",
      },
    },
    "message": "commit c:1, d:1",
    "type": "put",
  },
]
`;
